perf/gfoidl.Stochastics.Benchmarks/LoopSimdBenchmarks.cs

```diff
@@ -0,0 +1,391 @@
+
+                        vec = VectorHelper.GetVector(arr + 6 * Vector<double>.Count);
+                        res += vec;
+
+                        vec = VectorHelper.GetVector(arr + 7 * Vector<double>.Count);
+                        res += vec;
+
+                        arr += 8 * Vector<double>.Count;
+                    }
+
+                    for (; i < n - 4 * Vector<double>.Count; i += 4 * Vector<double>.Count)
+                    {
+                        Vector<double> vec = VectorHelper.GetVector(arr);
+                        res += vec;
+
+                        vec = VectorHelper.GetVector(arr + Vector<double>.Count);
+                        res += vec;
+
+                        vec = VectorHelper.GetVector(arr + 2 * Vector<double>.Count);
+                        res += vec;
+
+                        vec = VectorHelper.GetVector(arr + 3 * Vector<double>.Count);
+                        res += vec;
+
+                        arr += 4 * Vector<double>.Count;
+                    }
+
+                    for (; i < n - 2 * Vector<double>.Count; i += 2 * Vector<double>.Count)
+                    {
+                        Vector<double> vec = VectorHelper.GetVector(arr);
+                        res += vec;
+
+                        vec = VectorHelper.GetVector(arr + Vector<double>.Count);
+                        res += vec;
+
+                        arr += 2 * Vector<double>.Count;
+                    }
+
+                    _res = res;
+                }
+
+                for (; i < n; ++i)
+                    pArray[i] += 42d;
+            }
+
+            return _res[0];
+        }
+        //---------------------------------------------------------------------
+        [Benchmark]
+        public unsafe double Unrolled_8x2()
+        {
+            int n = _values.Length;
+
+            fixed (double* pArray = _values)
+            {
+                double* arr = pArray;
+                int i = 0;
+
+                if (Vector.IsHardwareAccelerated && n >= Vector<double>.Count * 2)
+                {
+                    Vector<double> res = _res;
+
+                    for (; i < n - 8 * Vector<double>.Count; i += 8 * Vector<double>.Count)
+                    {
+                        Core(arr, 0 * Vector<double>.Count, ref res);
+                        Core(arr, 1 * Vector<double>.Count, ref res);
+                        Core(arr, 2 * Vector<double>.Count, ref res);
+                        Core(arr, 3 * Vector<double>.Count, ref res);
+                        Core(arr, 4 * Vector<double>.Count, ref res);
+                        Core(arr, 5 * Vector<double>.Count, ref res);
+                        Core(arr, 6 * Vector<double>.Count, ref res);
+                        Core(arr, 7 * Vector<double>.Count, ref res);
+
+                        arr += 8 * Vector<double>.Count;
+                    }
+
+                    for (; i < n - 4 * Vector<double>.Count; i += 4 * Vector<double>.Count)
```

Comment:
These loops can be eliminated, becuase they will never be executed.

Unrolled 8x -> remainder max. 7
Unrolled 4x -> remainder max. 3
Unrolled 2x -> remainder max. 1

So, no loop is possible. Thus the loop reinitialization instructions can be saved, hence the jumps are nearer.

```asm
G_M10047_IG06:
       C4E17D1002           vmovupd  ymm0, ymmword ptr[rdx]
       C4E17558C8           vaddpd   ymm1, ymm0
       C4E17D104220         vmovupd  ymm0, ymmword ptr[rdx+32]
       C4E17558C8           vaddpd   ymm1, ymm0
       C4E17D104240         vmovupd  ymm0, ymmword ptr[rdx+64]
       C4E17558C8           vaddpd   ymm1, ymm0
       C4E17D104260         vmovupd  ymm0, ymmword ptr[rdx+96]
       C4E17558C8           vaddpd   ymm1, ymm0
       C4E17D108280000000   vmovupd  ymm0, ymmword ptr[rdx+128]
       C4E17558C8           vaddpd   ymm1, ymm0
       C4E17D1082A0000000   vmovupd  ymm0, ymmword ptr[rdx+160]
       C4E17558C8           vaddpd   ymm1, ymm0
       C4E17D1082C0000000   vmovupd  ymm0, ymmword ptr[rdx+192]
       C4E17558C8           vaddpd   ymm1, ymm0
       C4E17D1082E0000000   vmovupd  ymm0, ymmword ptr[rdx+224]
       C4E17558C8           vaddpd   ymm1, ymm0
       4881C200010000       add      rdx, 256
       83C120               add      ecx, 32
       413BC9               cmp      ecx, r9d
       7C8E                 jl       SHORT G_M10047_IG06

G_M10047_IG07:
       448D4EF0             lea      r9d, [rsi-16]
       413BC9               cmp      ecx, r9d
       7D35                 jge      SHORT G_M10047_IG08
       C4E17D1002           vmovupd  ymm0, ymmword ptr[rdx]
       C4E17558C8           vaddpd   ymm1, ymm0
       C4E17D104220         vmovupd  ymm0, ymmword ptr[rdx+32]
       C4E17558C8           vaddpd   ymm1, ymm0
       C4E17D104240         vmovupd  ymm0, ymmword ptr[rdx+64]
       C4E17558C8           vaddpd   ymm1, ymm0
       C4E17D104260         vmovupd  ymm0, ymmword ptr[rdx+96]
       C4E17558C8           vaddpd   ymm1, ymm0
       4881C280000000       add      rdx, 128
       83C110               add      ecx, 16

G_M10047_IG08:
       448D4EF8             lea      r9d, [rsi-8]
       413BC9               cmp      ecx, r9d
       7D18                 jge      SHORT G_M10047_IG09
       C4E17D1002           vmovupd  ymm0, ymmword ptr[rdx]
       C4E17558C8           vaddpd   ymm1, ymm0
       C4E17D104220         vmovupd  ymm0, ymmword ptr[rdx+32]
       C4E17558C8           vaddpd   ymm1, ymm0
       83C108               add      ecx, 8
```

Reply:
Is done in method `Unrolled_8x2`, not in this one for reference.